src/cpg_flow/workflow.py

```diff
@@ -286,75 +304,24 @@ def _process_first_last_stages(
         before first_stages, and all stages after last_stages (i.e. descendants and
         ancestors on the stages DAG.)
         """
-        stages_d = {s.name: s for s in stages}
-        stage_names = list(stg.name for stg in stages)
-        lower_names = {s.lower() for s in stage_names}
-
-        for param, _stage_list in [
-            ('first_stages', first_stages),
-            ('last_stages', last_stages),
-        ]:
-            for _s_name in _stage_list:
-                if _s_name.lower() not in lower_names:
-                    raise WorkflowError(
-                        f'Value in workflow/{param} "{_s_name}" must be a stage name '
-                        f'or a subset of stages from the available list: '
-                        f'{", ".join(stage_names)}',
-                    )
-
-        if not (last_stages or first_stages):
+        if not (first_stages or last_stages):
```

Comment:
I don't think it matters overly much, but this gatekeeping is already present (as in 'if not first or last, don't call this method'). This here as extra security?

Reply:
Good point. We shouldn't have both, and I would prefer to have it in the function because that's required for its correctness. Without the early out, it would skip everything, since there are no ancestors or descendants of the empty set of nodes, so `kept` would be empty.